### PR TITLE
EPOLL ET AutoRead

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
EPOLL does not support autoread when in ET mode.

Modifications:
- EpollRecvByteAllocatorHandle should not unconditionally force reading just because ET is enabled
- AbstractEpollChannel and all derived classes which implement epollInReady must support a variable which indicates
there may be more data to read. The variable will be used when read is called to simulate a EPOLL wakeup and call epollInReady if necessary. This will ensure that if we don't read until EAGAIN that we will try to read again and not rely on EPOLL to notify us.

Result:
EPOLL ET supports auto read.

### CODE REVIEW

**Comment by Scottmitch** [comment]
@normanmaurer @nmittler - PTAL.

@blucas - Please verify this works.

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollRecvByteAllocatorStreamingHandle.java [hunk 1]
```diff
@@ -15,27 +15,19 @@
...
 final class EpollRecvByteAllocatorStreamingHandle extends EpollRecvByteAllocatorHandle {
-    public EpollRecvByteAllocatorStreamingHandle(RecvByteBufAllocator.Handle handle, boolean isEdgeTriggered) {
-        super(handle, isEdgeTriggered);
+    public EpollRecvByteAllocatorStreamingHandle(RecvByteBufAllocator.Handle handle, ChannelConfig config) {
+        super(handle, config);
     }
 
-    @Override
-    public boolean continueReading() {
+    boolean maybeMoreDataToRead() {
```
Comment (netkins) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollRecvByteAllocatorStreamingHandle.java]: MAJOR Add the "@Override" annotation above this method signature [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1161)

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollRecvByteAllocatorHandle.java [hunk 2]
```diff
@@ -15,26 +15,47 @@
...
+         * EPOLL ET requires that we read until we get an EAGAIN
+         * (see Q9 in <a href="http://man7.org/linux/man-pages/man7/epoll.7.html">epoll man</a>). However in order to
+         * respect auto read we supporting reading to stop if auto read is off. If auto read is on we force reading to
+         * continue to avoid a {@link java.lang.StackOverflowError} between channelReadComplete and reading from the
+         * channel. It is expected that the {@link #EpollSocketChannel} implementations will track if we are in
+         * edgeTriggered mode and all data was not read, and will force a EPOLLIN ready event.
+         *
+         * If EPOLLRDHUP has been received we must read until we get a read error.
+         */
+        return receivedRdHup ? true : maybeMoreDataToRead() && config.isAutoRead() || super.continueReading();
```
Comment (netkins) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollRecvByteAllocatorHandle.java]: MAJOR Ternary operators that can be simplified with || or && [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3ASimplifiedTernary)

**Comment by normanmaurer** [comment]
@Scottmitch please fix the buffer leaks.

**Comment by Scottmitch** [comment]
@normanmaurer - Leak in unit test fixed.

**Comment by normanmaurer** [comment]
@Scottmitch LGTM...

@blucas @trustin PTAL.

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollRecvByteAllocatorHandle.java [hunk 3]
```diff
@@ -15,26 +15,47 @@
...
-    public final boolean isEdgeTriggered() {
-        return isEdgeTriggered;
+    final void receivedRdHup() {
+        receivedRdHup = true;
     }
 
-    public final void receivedRdHup() {
-        receivedRdHup = true;
+    boolean maybeMoreDataToRead() {
+        return isEdgeTriggered && lastBytesRead() > 0;
```
Comment (Scottmitch) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollRecvByteAllocatorHandle.java]: @trustin @normanmaurer - We may be able to do the same thing for LT mode too, but I figured lets just start with ET where it is necessary and add it as an optimization for LT later if needed.

**Comment by Scottmitch** [comment]
@blucas - any updates?

**Comment by blucas** [comment]
@Scottmitch - Sorry, I haven't had the chance to do a decent test with this branch. I should be able to do something tonight or tomorrow night.  I have looked at the code, and it seems okay to me.

**Comment by normanmaurer** [comment]
@blucas thanks please let us know once you had a chance.

**Comment by blucas** [comment]
@Scottmitch @normanmaurer - I've tested this using TCP traffic and it seems good.

**Comment by Scottmitch** [comment]
Cherry-picked 4.1 (bfbef03)

Unfortunately this is not a straight forward back port to 4.0 because of the `RecvByteAllocator` changes. I'm not planning on back porting this change ATM.

**Comment by normanmaurer** [comment]
No backport is fine

**Pull request closed without merging**